File: Assignment 3: Collinear Points/Python/BruteCollinearPoints.py
from Point import Point
from LineSegment import LineSegment

class BruteCollinearPoints:
    def __init__(self, points):
        """
        Initializes a BruteCollinearPoints object with an array of points.

        Args:
            points (list[Point]): An array of points.
        Raises:
            ValueError: If the input array or any point in it is None.
            ValueError: If any two points in the array are the same.
        """
        if points is None:
            raise ValueError()

        n = len(points)
        self.segments = []

        for p in range(n):
            if points[p] is None:
                raise ValueError()
            for q in range(p + 1, n):
                if points[q] is None or points[p].slopeTo(points[q]) == float("-inf"):
                    raise ValueError()
                for r in range(q + 1, n):
                    if points[r] is None or points[q].slopeTo(points[r]) == float("-inf"):
                        raise ValueError()
                    if points[p].slopeTo(points[q]) == points[p].slopeTo(points[r]):
                        for s in range(r + 1, n):
                            if points[s] is None or points[r].slopeTo(points[s]) == float("-inf"):
                                raise ValueError()
                            if points[p].slopeTo(points[q]) == points[p].slopeTo(points[s]):
                                segment = [points[p], points[q], points[r], points[s]]
                                segment.sort()
                                self.segments.append(LineSegment(segment[0], segment[3]))

    def number_of_segments(self):
        """
        Returns the number of line segments found.

        Returns:
            int: The number of line segments.
        """
        return len(self.segments)

    def get_segments(self):
        """
        Returns an array of line segments found.

        Returns:
            list[LineSegment]: An array of line segments.
        """
        return self.segments


# Example usage
if __name__ == "__main__":
    points = [Point(10000, 0), Point(0, 10000), Point(3000, 7000), Point(7000, 3000), Point(20000, 21000),
        	  Point(3000, 4000), Point(14000, 15000), Point(6000, 7000)]
    
    lines = BruteCollinearPoints(points)
    print(f"Number of segments found = {lines.number_of_segments()}")

    segments = lines.get_segments()
    for segment in segments:
    	print(segment)

File: Assignment 3: Collinear Points/Java/BruteCollinearPoints.java
import edu.princeton.cs.algs4.In;
import edu.princeton.cs.algs4.StdDraw;
import edu.princeton.cs.algs4.StdOut;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class BruteCollinearPoints {
    private List<LineSegment> segments = new ArrayList<>();

    /**
     * Finds all line segments containing 4 points using the brute-force approach.
     * Checks all combinations of 4 points and identifies collinear segments.
     *
     * @param points an array of points
     * @throws IllegalArgumentException if the input array or any point in it is null
     */
    public BruteCollinearPoints(Point[] points) {
        if (points == null) {
            throw new IllegalArgumentException();
        }

        int n = points.length;

        for (int p = 0; p < n; p++) {
            if (points[p] == null) throw new IllegalArgumentException();
            for (int q = p + 1; q < n; q++) {
                if (points[q] == null || points[p].slopeTo(points[q]) == Double.NEGATIVE_INFINITY) {
                    throw new IllegalArgumentException();
                }
                for (int r = q + 1; r < n; r++) {
                    if (points[r] == null
                            || points[q].slopeTo(points[r]) == Double.NEGATIVE_INFINITY) {
                        throw new IllegalArgumentException();
                    }
                    if (points[p].slopeTo(points[q]) == points[p].slopeTo(points[r])) {
                        for (int s = r + 1; s < n; s++) {
                            if (points[s] == null
                                    || points[r].slopeTo(points[s]) == Double.NEGATIVE_INFINITY) {
                                throw new IllegalArgumentException();
                            }
                            if (points[p].slopeTo(points[q]) == points[p].slopeTo(points[s])) {
                                Point[] segment = { points[p], points[q], points[r], points[s] };
                                Arrays.sort(segment);
                                segments.add(new LineSegment(segment[0], segment[3]));
                            }
                        }
                    }
                }
            }
        }
    }

    /**
     * Returns the number of line segments found.
     *
     * @return the number of line segments
     */
    public int numberOfSegments() {
        return segments.size();
    }

    /**
     * Returns an array of line segments found.
     *
     * @return an array of line segments
     */
    public LineSegment[] segments() {
        return segments.toArray(new LineSegment[0]);
    }

    /**
     * Reads points from a file, finds collinear segments, and displays them.
     *
     * @param args command-line arguments (input file containing points)
     */
    public static void main(String[] args) {

        // read the n points from a file
        In in = new In(args[0]);
        int n = in.readInt();
        Point[] points = new Point[n];
        for (int i = 0; i < n; i++) {
            int x = in.readInt();
            int y = in.readInt();
            points[i] = new Point(x, y);
        }

        // draw the points
        StdDraw.enableDoubleBuffering();
        StdDraw.setXscale(0, 32768);
        StdDraw.setYscale(0, 32768);
        for (Point p : points) {
            p.draw();
        }
        StdDraw.show();

        // print and draw the line segments
        FastCollinearPoints collinear = new FastCollinearPoints(points);
        for (LineSegment segment : collinear.segments()) {
            StdOut.println(segment);
            segment.draw();
        }
        StdDraw.show();
    }
}
